File: sqlparser/src/sql/SqlStatement.java
//@
package cdata.sql;
//@
/*#
using RSSBus.core;
using RSSBus;

namespace CData.Sql {
#*/

import core.ParserCore;
import rssbus.oputils.common.Utilities;

public abstract class SqlStatement implements ISqlElement, ISqlCloneable {
  protected Dialect dialectProcessor;
  protected SqlCollection<String> comments = new SqlCollection<String>();
  protected String tableName;
  protected SqlTable table;
  protected SqlCollection<SqlValueExpression> parasList = new SqlCollection<SqlValueExpression>();
  protected int parentParasNumber = 0;
  protected SqlCollection<SqlTable> _withClause = null;

  protected SqlStatement(Dialect dialectProcessor) {
    this.dialectProcessor = dialectProcessor;
  }

  public abstract void accept(ISqlQueryVisitor visitor) throws Exception;

  public Dialect getDialectProcessor() {
    return this.dialectProcessor;
  }

  public /*#virtual#*/ SqlTable getTable() {
    return table;
  }

  public SqlCollection<String> getComments() {
    return this.comments;
  }

  public /*#virtual#*/ SqlCollection<SqlColumn> getColumns() {
    return new SqlCollection<SqlColumn>();
  }

  public /*#virtual#*/ void setColumns(SqlCollection<SqlColumn> columns) {
  }

  public /*#virtual#*/ void setWithClause(SqlCollection<SqlTable> withes) {
    this._withClause = withes;
  }

  public /*#virtual#*/ SqlCollection<SqlTable> getWithClause() {return this._withClause;}

  public /*#virtual#*/ SqlConditionNode getCriteria() {
    return null;
  }

  public /*#virtual #*/void setCriteria(SqlConditionNode cond) {
  }

  public /*#virtual #*/void addCondition(SqlConditionNode cond) {
  }

  public SqlCollection<SqlColumn> getColumns(String name, String alias) {
    SqlCollection<SqlColumn> filteredIn = new SqlCollection<SqlColumn>();
    SqlCollection<SqlColumn> exps = getColumns();
    if (exps != null) {
      for (SqlColumn exp : exps) {
        boolean match = false;
        if (name != null) {
          match |= exp.equals(name);
        }
        if (alias != null) {
          String expAlias = exp.getAlias();
          match |=  (alias.equalsIgnoreCase(expAlias));
        }
        if (match) {
          filteredIn.add(exp);
        }
      }
    }
    return filteredIn;
  }

  public SqlCollection<String> GetColumnNames() {
    SqlCollection<String> colNames = new SqlCollection<String>();
    for (SqlColumn col : this.getColumns()) {
      colNames.add(col.getColumnName());
    }
    return colNames;
  }

  public SqlColumn getResolvedColumn(String tableOrAlias, String nameOrAlias) {
    if (nameOrAlias == null) {
      return null;
    }
    SqlTable resolvedTable = null;
    if (tableOrAlias != null && tableOrAlias.length() > 0) {
      resolvedTable = getResolvedTable(tableOrAlias);
      if (resolvedTable != null && resolvedTable.isNestedQueryTable()) {
        SqlQueryStatement nestedQuery = resolvedTable.getQuery();
        return nestedQuery.getResolvedColumn(null, nameOrAlias);
      }
    }
    for (SqlColumn column : this.getColumns()) {
      String name = column.getColumnName();
      String alias = column.getAlias();
      SqlTable t = column.getTable();
      boolean tableMatch = false;
      if (resolvedTable != null) {
        String resolvedName = resolvedTable.getName();
        String resolvedAlias = resolvedTable.getAlias();
        if (t != null) {
          if (tableOrAlias.equals(t.getName()) || tableOrAlias.equals(t.getAlias())) {
            tableMatch = true;
          } else if (resolvedName.equals(t.getName()) || resolvedName.equals(t.getAlias())) {
            tableMatch = true;
          } else if (resolvedAlias.equals(t.getName()) || resolvedAlias.equals(t.getAlias())) {
            tableMatch = true;
          } else {
            tableMatch = false;
          }
        } else {
          tableMatch = true;
        }
      } else {
        tableMatch = true;
      }
      if (!tableMatch) {
        continue;
      }
      if (nameOrAlias.equalsIgnoreCase(name) || nameOrAlias.equalsIgnoreCase(alias)) {
        return (SqlColumn)column.clone();
      }
    }
    return null;
  }

  public String getResolvedTableName(String nameOralias) {
    SqlTable t = getResolvedTable(nameOralias);
    if (t != null) {
      if (t.getQuery() != null) {
        SqlTable subQueryTable = t.getQuery().getTable();
        if (subQueryTable != null) {
          return t.getQuery().getResolvedTableName(subQueryTable.getName());
        }
      } else {
        return t.getName();
      }
    }

    return nameOralias;
  }

  public /*#virtual#*/  SqlTable getResolvedTable(String nameOralias) {
    SqlTable resolved = null;
    if (getTable() != null && nameOralias != null) {
      SqlTable t = getTable();
      if (Utilities.equalIgnoreCase(nameOralias, t.getName()) ||
         Utilities.equalIgnoreCase(nameOralias, t.getAlias())) {
        resolved = t;
      }
    } else {
      resolved = getTable();
    }
    return resolved;
  }

  public SqlCollection<SqlValueExpression> getParameterList() {
    return parasList;
  }

  public int getParentParasNumber() {
    return this.parentParasNumber;
  }

  public String getTableName() {
    if (!ParserCore.isValidTableName(tableName)) {
      if (getTable() != null) {
        SqlTable t = getTable().resolve();
        if (t != null && t.getName() != null) {
          return t.getName();
        } else {
          return "";
        }
      } else {
        return "";
      }
    } else {
      return tableName;
    }
  }

  public /*#virtual #*/void setTable(SqlTable t) {
    this.table = t;
  }

  public void setTableName(String name) {
    this.tableName = name;
  }

  public boolean isJoinQuery() {
    if (this instanceof SqlSelectStatement) {
      return ParserCore.isJoinStatement((SqlSelectStatement)this);
    }

    return false;
  }

  public abstract Object clone();

  protected /*#virtual #*/void copy(SqlStatement obj) {
    this.dialectProcessor = obj.dialectProcessor;
    this.comments = obj.comments == null ? null : (SqlCollection<String>)obj.comments.clone();
    this.tableName = obj.tableName;
    this.table = obj.table;
    this._withClause = obj._withClause;
    if (obj.parasList == null) {
      this.parasList = null;
    } else {
      this.parasList =  new SqlCollection<SqlValueExpression>();
      for (SqlValueExpression p : obj.parasList) {
        this.parasList.add(p);
      }
    }
  }

  public String toString() {
    try {
      SqlBuilder bld = SqlBuilder.createBuilder();
      bld.setStatement(this);
      return bld.build();
    } catch(Exception ex) {
      return Utilities.getExceptionMessage(ex);
    }
  }

}

File: sqlparser/src/sql/csgen/SqlStatement.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Net;
using System.Net.Sockets;
using System.Text;
using RSSBus.core.j2cs;



using RSSBus.core;
using RSSBus;

namespace CData.Sql {


public abstract class SqlStatement : ISqlElement, ISqlCloneable {
  protected Dialect dialectProcessor;
  protected SqlCollection<string> comments = new SqlCollection<string>();
  protected string tableName;
  protected SqlTable table;
  protected SqlCollection<SqlValueExpression> parasList = new SqlCollection<SqlValueExpression>();
  protected int parentParasNumber = 0;
  protected SqlCollection<SqlTable> _withClause = null;

  protected SqlStatement(Dialect dialectProcessor) {
    this.dialectProcessor = dialectProcessor;
  }

  public abstract void Accept(ISqlQueryVisitor visitor) ;

  public Dialect GetDialectProcessor() {
    return this.dialectProcessor;
  }

  public virtual SqlTable GetTable() {
    return table;
  }

  public SqlCollection<string> GetComments() {
    return this.comments;
  }

  public virtual SqlCollection<SqlColumn> GetColumns() {
    return new SqlCollection<SqlColumn>();
  }

  public virtual void SetColumns(SqlCollection<SqlColumn> columns) {
  }

  public virtual void SetWithClause(SqlCollection<SqlTable> withes) {
    this._withClause = withes;
  }

  public virtual SqlCollection<SqlTable> GetWithClause() {return this._withClause;}

  public virtual SqlConditionNode GetCriteria() {
    return null;
  }

  public virtual void SetCriteria(SqlConditionNode cond) {
  }

  public virtual void AddCondition(SqlConditionNode cond) {
  }

  public SqlCollection<SqlColumn> GetColumns(string name, string alias) {
    SqlCollection<SqlColumn> filteredIn = new SqlCollection<SqlColumn>();
    SqlCollection<SqlColumn> exps = GetColumns();
    if (exps != null) {
      foreach(SqlColumn exp in exps) {
        bool match = false;
        if (name != null) {
          match |= exp.Equals(name);
        }
        if (alias != null) {
          string expAlias = exp.GetAlias();
          match |=  (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(alias, expAlias));
        }
        if (match) {
          filteredIn.Add(exp);
        }
      }
    }
    return filteredIn;
  }

  public SqlCollection<string> GetColumnNames() {
    SqlCollection<string> colNames = new SqlCollection<string>();
    foreach(SqlColumn col in this.GetColumns()) {
      colNames.Add(col.GetColumnName());
    }
    return colNames;
  }

  public SqlColumn GetResolvedColumn(string tableOrAlias, string nameOrAlias) {
    if (nameOrAlias == null) {
      return null;
    }
    SqlTable resolvedTable = null;
    if (tableOrAlias != null && tableOrAlias.Length > 0) {
      resolvedTable = GetResolvedTable(tableOrAlias);
      if (resolvedTable != null && resolvedTable.IsNestedQueryTable()) {
        SqlQueryStatement nestedQuery = resolvedTable.GetQuery();
        return nestedQuery.GetResolvedColumn(null, nameOrAlias);
      }
    }
    foreach(SqlColumn column in this.GetColumns()) {
      string name = column.GetColumnName();
      string alias = column.GetAlias();
      SqlTable t = column.GetTable();
      bool tableMatch = false;
      if (resolvedTable != null) {
        string resolvedName = resolvedTable.GetName();
        string resolvedAlias = resolvedTable.GetAlias();
        if (t != null) {
          if (tableOrAlias.Equals(t.GetName()) || tableOrAlias.Equals(t.GetAlias())) {
            tableMatch = true;
          } else if (resolvedName.Equals(t.GetName()) || resolvedName.Equals(t.GetAlias())) {
            tableMatch = true;
          } else if (resolvedAlias.Equals(t.GetName()) || resolvedAlias.Equals(t.GetAlias())) {
            tableMatch = true;
          } else {
            tableMatch = false;
          }
        } else {
          tableMatch = true;
        }
      } else {
        tableMatch = true;
      }
      if (!tableMatch) {
        continue;
      }
      if (RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(nameOrAlias, name) || RSSBus.core.j2cs.Converter.GetEqualsIgnoreCase(nameOrAlias, alias)) {
        return (SqlColumn)column.Clone();
      }
    }
    return null;
  }

  public string GetResolvedTableName(string nameOralias) {
    SqlTable t = GetResolvedTable(nameOralias);
    if (t != null) {
      if (t.GetQuery() != null) {
        SqlTable subQueryTable = t.GetQuery().GetTable();
        if (subQueryTable != null) {
          return t.GetQuery().GetResolvedTableName(subQueryTable.GetName());
        }
      } else {
        return t.GetName();
      }
    }

    return nameOralias;
  }

  public virtual  SqlTable GetResolvedTable(string nameOralias) {
    SqlTable resolved = null;
    if (GetTable() != null && nameOralias != null) {
      SqlTable t = GetTable();
      if (Utilities.EqualIgnoreCase(nameOralias, t.GetName()) ||
         Utilities.EqualIgnoreCase(nameOralias, t.GetAlias())) {
        resolved = t;
      }
    } else {
      resolved = GetTable();
    }
    return resolved;
  }

  public SqlCollection<SqlValueExpression> GetParameterList() {
    return parasList;
  }

  public int GetParentParasNumber() {
    return this.parentParasNumber;
  }

  public string GetTableName() {
    if (!ParserCore.IsValidTableName(tableName)) {
      if (GetTable() != null) {
        SqlTable t = GetTable().Resolve();
        if (t != null && t.GetName() != null) {
          return t.GetName();
        } else {
          return "";
        }
      } else {
        return "";
      }
    } else {
      return tableName;
    }
  }

  public virtual void SetTable(SqlTable t) {
    this.table = t;
  }

  public void SetTableName(string name) {
    this.tableName = name;
  }

  public bool IsJoinQuery() {
    if (this is SqlSelectStatement) {
      return ParserCore.IsJoinStatement((SqlSelectStatement)this);
    }

    return false;
  }

  public abstract Object Clone();

  protected virtual void Copy(SqlStatement obj) {
    this.dialectProcessor = obj.dialectProcessor;
    this.comments = obj.comments == null ? null : (SqlCollection<string>)obj.comments.Clone();
    this.tableName = obj.tableName;
    this.table = obj.table;
    this._withClause = obj._withClause;
    if (obj.parasList == null) {
      this.parasList = null;
    } else {
      this.parasList =  new SqlCollection<SqlValueExpression>();
      foreach(SqlValueExpression p in obj.parasList) {
        this.parasList.Add(p);
      }
    }
  }

  public string ToString() {
    try {
      SqlBuilder bld = SqlBuilder.CreateBuilder();
      bld.SetStatement(this);
      return bld.Build();
    } catch(Exception ex) {
      return Utilities.GetExceptionMessage(ex);
    }
  }

}
}

